Add numeric output to version command.

pgstef opened the pull request:
Hi :wave: 

This is a WIP ;-)

- Added `info --output=json` fields `"backrest":{"version":"2.55dev","version-num":"025500"}`
- Not really knowing where it would be best to place it, the convert function is still is the info command
- Tried adding a `version --output=num` option to return the same and use the above function

At the moment, I can build the code manually but CI seems to fail on `error: ‘cfgOptOutput’ undeclared`. Not sure what I missed regarding the help/version command :-(

dwsteele commented:
>  I can build the code manually but CI seems to fail on error: ‘cfgOptOutput’ undeclared

This failure happened when building `test-pgbackrest`. The test and doc code also use the help/version code so they need the options added as well.

I also simplified the code by adding a constant for version num.

I'm not sure I'm entirely happy with the info output, but go ahead and fix the coverage for the help module while I think about it.

pgstef commented on src/version.h:
```diff
@@ -31,13 +31,14 @@ will be invalid unless migration functions are written.
 #define PROJECT_VERSION_SUFFIX                                      "dev"
 
 #define PROJECT_VERSION                                             "2.55.0dev"
+#define PROJECT_VERSION_NUM                                         2055000
```
you're counting 1 digit for major, and 3 digits for minor and patch then?

dwsteele replied:
Yes -- for future proofing. We are already at 2.55.

dwsteele replied:
Also, there are many digits as we like to major version -- we just don't prepend 0s. You'll notice that I also changed version-num to a number in the JSON. This makes comparisons easier and is equivalent to what Postgres does.

pgstef replied:
Cool. Thanks for the clarification :+1:

pgstef commented:
Thanks for having a look at this :pray: 

> This failure happened when building `test-pgbackrest`. The test and doc code also use the help/version code so they need the options added as well.

I remember I tried adding it to the doc config.yaml/help.xml but not to the help one :facepalm:

> I also simplified the code by adding a constant for version num.

That was my initial idea but I wasn't sure you'd want to keep that up to date so I went for the extract function. But obviously, that makes it much simpler with the constant.

> I'm not sure I'm entirely happy with the info output, but go ahead and fix the coverage for the help module while I think about it.

As the json output loops over each stanza, it felt weird to me to have the same output in all the stanzas. That's why I went for the version option initially. Monitoring the version installed on your hosts separately with the version command seemed more reliable to me. But as the initial request was to add to the json info, I guess having it all in one command might feel easier to handle.

Regarding the help test, I tried adding something simple ([31ac9f4](https://github.com/pgbackrest/pgbackrest/pull/2526/commits/31ac9f4493f996fe3ad349412589c6fd21e979fc)) but that broke more tests with `BUT GOT AssertError: expected return type 'String' already found but also found 'String'` so I reverted it and will need to have a closer look at it next week :sweat_smile:

dwsteele commented:
> That was my initial idea but I wasn't sure you'd want to keep that up to date so I went for the extract function.

This is automatically updated in test.pl so there is nothing to maintain.

> As the json output loops over each stanza, it felt weird to me to have the same output in all the stanzas.

Same here, but I can't see anywhere else to put it.

> AssertError: expected return type 'String' already found but also found 'String'

This is because the strNewFmt outside of the MEM_CONTEXT_TEMP block was leaking a string. Fixed that.

pgstef commented:
> This is because the strNewFmt outside of the MEM_CONTEXT_TEMP block was leaking a string. 

Oh, okay. Good to spot it with the tests then. That's again a good reason to add tests :laughing: 

> Fixed that.

Thanks again for your help on this PR :pray: I definitely learned a lot on this :bowing_man:

pgstef commented:
> Same here, but I can't see anywhere else to put it.

First of all, I must admit I'm pretty happy with `version --output=num`. I believe it will be very efficient to monitor the version installed on the various hosts used and may help detect potential mismatch earlier and easier.

Although, I've read again the initial [issue](https://github.com/pgbackrest/pgbackrest/issues/2410) and given what @keithf4 describes there and having a look at the pgmonitor [queries](https://github.com/CrunchyData/pgmonitor/blob/main/postgres_exporter/linux/queries_backrest.yml#L23), I feel like they're currently extracting `backrest_repo_version` from the `backup:backrest:version` json key. That's then the version which was used to take the backup, not the version currently installed. As we have the limitation on having same version on PG+repo, "repo version" sounds accurate enough.

I'm now wondering if we shouldn't consider to add the version num to `InfoBackupData` too :thinking:
With that value, we could then compare the backup version with the currently installed version. I would then ask if comparing against `version --output=num` would be easy for them to add or if we should keep the new info json field too.

That's maybe a bigger change but I kinda feel it would worth it :man_shrugging: What do you think?

dwsteele commented:
> First of all, I must admit I'm pretty happy with version --output=num. I believe it will be very efficient to monitor the version installed on the various hosts used and may help detect potential mismatch earlier and easier.

Same here. As far as I'm concerned that is ready to commit.

I'm a little less happy with the way the version is expressed in info per stanza but I don't see that we have any choice.

> I would then ask if comparing against version --output=num would be easy for them to add or if we should keep the new info json field too.

I pinged @keithf4 about it, so hopefully we'll find out.

> I'm now wondering if we shouldn't consider to add the version num to InfoBackupData too

That's a good idea. I'd actually want to add it to the manifest at backup time so we don't try to go back and translate old version numbers. I worried about a problem there that would cause errors (especially in forks). I think having version_num for backups going forward is fine.

pgstef commented:
> I'm a little less happy with the way the version is expressed in info per stanza but I don't see that we have any choice.

> I'd actually want to add it to the manifest at backup time so we don't try to go back and translate old version numbers. I worried about a problem there that would cause errors (especially in forks). I think having version_num for backups going forward is fine.

I believe adding the version in the backup manifest for future backups and report that in the per backup json field rather than per stanza would make sense. So you know what version was use to generate the backup and use the version command to get the locally running version to compare both info :thinking:

pgstef commented:
> I think we need both -- the version when a backup was made _and_ the current version. The latter needs to be per stanza due to limitations of the format.

Attempted to add it in 38cb4cc. Hopefully the test should run fine :crossed_fingers:

pgstef commented:
> I've been thinking about this and for now I'd rather just go with the --output=num feature for the version command. I'm ready to commit that right now, but still not satisfied with the rest. Not your fault -- I'm just uncomfortable with how this has to be jammed in given the current format.

Same feeling here. I like the `version --output=num` too :+1: 

> Could you update this PR to just have that initial feature? Keep your original branch as well because we may come back to that in time.

Ok, I'll try to do that asap

dwsteele reviewed (approved):
> Except for the arch (--platform=linux/ppc64le ubuntu:22.04) test which fails for https://github.com/pgbackrest/pgbackrest/issues/2568 reason

This has been fixed and merged into the commit.

The rest looks good to me and is ready to commit. Thank you!
